pkg/kf/logs/tailer.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+// Tailer reads the logs for a KF application. It should be created via
+// NewTailer().
+type Tailer interface {
+	// Tail tails the logs from a KF application and writes them to the
+	// writer.
+	Tail(ctx context.Context, appName string, out io.Writer, opts ...TailOption) error
+}
+
+type tailer struct {
+	client corev1.CoreV1Interface
+}
+
+// NewTailer creates a new Tailer.
+func NewTailer(client corev1.CoreV1Interface) Tailer {
+	return &tailer{
+		client: client,
+	}
+}
+
+// Tail tails the logs from a KF application and writes them to the writer.
+func (t *tailer) Tail(ctx context.Context, appName string, out io.Writer, opts ...TailOption) error {
+	cfg := TailOptionDefaults().Extend(opts).toConfig()
+	if appName == "" {
+		return errors.New("appName is empty")
+	}
+
+	if cfg.NumberLines < 0 {
+		return errors.New("number of lines must be greater than or equal to 0")
+	}
+
+	namespace := cfg.Namespace
+	logOpts := v1.PodLogOptions{
+		Container: "user-container",
```

Comment:
documenting this magic string would be nice.

Reply:
agreed, because I do not understand what exactly this is doing

Reply:
Done